src/util.c

```diff
@@ -132,44 +132,303 @@ long PURE str_to_long(const char *str)
+        count += bits; // [1, 9, 17, 25]
+    }
+    if (1) {
+        // iteration 2
+        mask >>= (1 << 2); // 0x0000000f
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 2; // [0, 4]
+        x >>= bits; // <= 0x0000000f
+        count += bits; // [1, 5, 9, ..., 29]
     }
-    return count;
+    if (1) {
+        // iteration 1
+        mask >>= (1 << 1); // 0x00000003
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 1; // [0, 2]
+        x >>= bits; // <= 0x00000003
+        count += bits; // [1, 3, 5, ..., 31]
+    }
+    if (1) {
+        // iteration 0
+        mask >>= (1 << 0); // 0x00000001
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 0; // [0, 1]
+        x >>= bits; // <= 0x00000001
+        count += bits; // [1, 2, 3, ..., 32]
+    }
+
+    // If the original input was zero, there will have been no shifts, so this
+    // gives a result of 32. Otherwise, x is now exactly 1, so subtracting from
+    // count gives a result from [0, 1, 2, ..., 31].
+    return count - x;
 }
 
-uint32_t __ctzdi2(uint64_t x)
+static inline unsigned ctz64(uint64_t x)
 {
-    uint32_t count = 0;
-    while (!(x & 0x00000000000000001) && count <= 64) {
-        x >>= 1;
-        count++;
+    unsigned count = 1;
+    uint64_t mask = UINT64_MAX;
+
+    if (1) {
+        // iteration 5
+        mask >>= (1 << 5); // 0x00000000ffffffff
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 5; // [0, 32]
+        x >>= bits; // <= 0x00000000ffffffff
+        count += bits; // [1, 33]
+    }
+    if (1) {
+        // iteration 4
+        mask >>= (1 << 4); // 0x000000000000ffff
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 4; // [0, 16]
+        x >>= bits; // <= 0x000000000000ffff
+        count += bits; // [1, 17, 33, 49]
     }
-    return count;
+    if (1) {
+        // iteration 3
+        mask >>= (1 << 3); // 0x00000000000000ff
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 3; // [0, 8]
+        x >>= bits; // <= 0x00000000000000ff
+        count += bits; // [1, 9, 17, ..., 57]
+    }
+    if (1) {
+        // iteration 2
+        mask >>= (1 << 2); // 0x000000000000000f
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 2; // [0, 4]
+        x >>= bits; // <= 0x000000000000000f
+        count += bits; // [1, 5, 9, ..., 61]
+    }
+    if (1) {
+        // iteration 1
+        mask >>= (1 << 1); // 0x0000000000000003
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 1; // [0, 2]
+        x >>= bits; // <= 0x0000000000000003
+        count += bits; // [1, 3, 5, ..., 63]
+    }
+    if (1) {
+        // iteration 0
+        mask >>= (1 << 0); // 0x0000000000000001
+        unsigned bits = ((unsigned)((x & mask) == 0)) << 0; // [0, 1]
+        x >>= bits; // <= 0x0000000000000001
+        count += bits; // [1, 2, 3, ..., 64]
+    }
+
+    return count - x;
+}
+
+unsigned ctzl_impl(unsigned long x)
+{
+    return sizeof(unsigned long) == 8 ? ctz64(x) : ctz32(x);
```

Comment:
this can be determined during compiling time, so the ? operator could be saved?

Reply:
Yeah, this is just to make it somewhat portable, so `ctzl` will work on platforms where `unsigned long` is either 32 or 64 bits. But it is a const expression, so it compiles away. Apparently, you can't use `sizeof` in a `#if`, so this seemed the clearest way to do it.